Rebased CBOR Changes

tobiasmo1 opened the pull request:
Resubmitting changes with signed commits and rebased to latest 'master'.

These squashed commits have been rebased on top of AutoEvents PR (#216).

The go-codec package uses unsafe for performance. To reduce possible point of failure we will build with 'safe' flag to avoid this optimization. We plan to revisit and to remove this build flag when performance and functional validation tests are in place.

The following functionality is _not_ included:

- Handling of PUT commands using binary values.
- Assignment of MediaType to Event readings
- Avoid duplicate encoding performed for Event and Response. Will be addressed when EventClient.Add returns payload as byte array for use in command response.
- Retrieval of global configured limit of MaxBinaryBytes. May need to reside in Consul (tbd).
- Logging of application errors from CommandHandler (will add defect).
- Error handling of example and the current use of multiple media types from a single DeviceCommand; since support is currently limited to a single ValueDescriptor per DeviceCommand.

These will be addressed separately from this PR.

iain-anderson commented on pkg/models/commandvalue.go:
```diff
@@ -81,97 +91,115 @@ type CommandValue struct {
+	switch t {
+		case Binary:
+			// assign cv.BinValue
+			err = encodeBinaryValue(cv, value)
```
Do we need to encode here, or is the encoding of Events sufficient?

tobiasmo1 replied:
Encoding of Events should be sufficient.
Will be addressed in a follow up PR.

tonyespy replied:
So after a close look, it appears the code is assuming that all parameters passed into to a write command are decoded as JSON, and then this function seems to encoding binary values into CBOR before being sent to the ProtocolDriver via handleWriteCommands().

I think what needs to happen is on a write command from a client, if one or more parameters are binary, then the whole request body should be CBOR encoded. So the write command logic should check the content header and decoded appropriately. The resulting binary data should be passed to the driver as a byte array.

Maybe it makes sense for handling write commands for this PR?

tobiasmo1 replied:
@tonyespy That makes sense. I will add handling write commands to a new PR.

tsconn23 commented:
recheck

tobiasmo1 commented:
@iain-anderson @cloudxxx8 @tsconn23 @tonyespy 
I rebased CBOR changes onto latest master, resolved conflicts and adjusted the AutoEvents to use the local DSDK event model.

I confirmed the runtime behavior of AutoEvents and responses to GET of boolean (Switch) and binary (Image) DeviceCommands initiated w/ Postman.

tsconn23 commented:
@tobiasmo1 With @cloudxxx8 approval please squash commits so we can get this merged.

tobiasmo1 commented:
> @tobiasmo1 With @cloudxxx8 approval please squash commits so we can get this merged.

Done.

tobiasmo1 commented:
Done.

From: Trevor Conn [mailto:notifications@github.com]
Sent: Wednesday, April 17, 2019 7:07 AM
To: edgexfoundry/device-sdk-go <device-sdk-go@noreply.github.com>
Cc: Mosby, Tobias <tobias.mosby@intel.com>; Mention <mention@noreply.github.com>
Subject: Re: [edgexfoundry/device-sdk-go] Rebased CBOR Changes (#243)


@tobiasmo1<https://github.com/tobiasmo1> With @cloudxxx8<https://github.com/cloudxxx8> approval please squash commits so we can get this merged.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/edgexfoundry/device-sdk-go/pull/243#issuecomment-484105201>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEAJEVMVRKMANWSXONRE273PQ4VKBANCNFSM4HF43UDQ>.

tobiasmo1 commented:
Fyi - Each time I push rebased commits it resets review approval requirements.
The GitHub UI provides a "Squash and Merge" button that avoids impacting review approvals.

I believe this repository is already [configured](https://help.github.com/en/articles/configuring-commit-squashing-for-pull-requests) to allow commit squashing, so we could expedite merges this way.

tsconn23 commented on go.mod:
```diff
@@ -9,5 +9,6 @@ require (
 	github.com/pelletier/go-toml v1.2.0
 	github.com/pkg/errors v0.8.1
 	github.com/stretchr/testify v1.3.0
+	github.com/ugorji/go/codec v0.0.0-20190316192920-e2bddce071ad
```
@tonyespy I wasn't concerned by it b/c we're on Go 1.11.x. My takeaway is that we don't need to apply a "safe" flag beyond 1.9.

The only thing I might add is that an actual release of this library (v1.1.4) came out on 8-Apr. I might prefer to pin to that than the commit above.

https://github.com/ugorji/go/releases/tag/v1.1.4

tobiasmo1 replied:
Yes, it is important to understand that go-codec does use the 'unsafe' package, and the implications of that. https://go101.org/article/unsafe.html.

I agree on pinning and will adjust go.mod to reference the release version.

tonyespy replied:
My interpretation was that they only support 3 versions via unsafe: 1.7, 1.8, and 1.9. My vote would be to use "safe" mode instead. If we find that it's a huge hit in performance, then we can consider re-building in unsafe mode.

tobiasmo1 replied:
@tonyespy Will either update Makefile to build with "safe" or I will update the PR.

tobiasmo1 commented on internal/provision/profiles.go:
```diff
@@ -124,7 +124,7 @@ func createDescriptorFromResourceOperation(profileName string, op models.Resourc
 		dr, ok := cache.Profiles().DeviceResource(profileName, op.Object)
 		if !ok {
-			common.LoggingClient.Error(fmt.Sprintf("can't find Device Object %s to match Resource Operation %v in Device Profile %s", op.Object, op, profileName))
+			common.LoggingClient.Error(fmt.Sprintf("can't find Device Resource %s to match Device Command (Resource Operation) %v in Device Profile %s", op.Object, op, profileName))
```
I added https://github.com/edgexfoundry/device-sdk-go/issues/245 for review/triage.

tobiasmo1 commented on example/driver/simpledriver.go:
```diff
@@ -25,6 +30,38 @@ type SimpleDriver struct {
+	}
+	defer img.Close()
+
+	// TODO: Attach MediaType property, determine if decoding
```
It may save cycles to just add it to this PR then.
My intent was to foster a conversation via code/PR (treating master as a typical feature branch) since that seemed to be the operating model. But I'll adjust.

tonyespy replied:
No need to add, just please add an explanation to the PR description. The idea is to provide as much context and information in order facilitate review of the PR. @anonymouse64 writes some really good PR descriptions for snap changes that many times include how to test the PR.

tobiasmo1 replied:
:thumbsup: Got it. Okay will address this in a separate PR and updated this PR's description.

tobiasmo1 commented on pkg/models/commandvalue.go:
```diff
@@ -81,97 +91,115 @@ type CommandValue struct {
+	// capacity of 1MB, used to hold binary values returned
+	// by a ProtocolDriver instance. Its decoded value is externally accessed
+	// using BinaryValue() method
+	BinValue []byte
```
We intentionally hold binary values and numbers in different fields; especially when I saw that separate encoding is already being performed on NumericValue. What is the purpose of the current BigEndian byte encoding performed on NumericValue?

tobiasmo1 replied:
We discussed that keeping it separate is acceptable.

tobiasmo1 commented on internal/common/utils.go:
```diff
@@ -46,9 +51,29 @@ func CommandValueToReading(cv *ds_models.CommandValue, devName string) *models.R
+	//  Alternately, EventClient will support a more generic interface; e.g., accept encoded event []byte and
+	//  allow caller to communicate needed details via Context (such as ContentType, etc.).
+	// _, err := EventClient.Add(&event.EncodedEvent, ctx)
+	_, err := EventClient.Add(&event.Event, ctx)
```
As we discussed the Add method will need to return bytes. So we will revisit when that lands.

tobiasmo1 commented on internal/autoevent/executor.go:
```diff
@@ -64,7 +70,7 @@ func (e *executor) Run() {
 }
 
-func readResource(e *executor) (*models.Event, common.AppError) {
+func readResource(e *executor) (*ds_models.Event, common.AppError) {
```
As we discussed, this was the only of these methods that needed the ds_models version for AutoEvents to prepare/submit to SendEvent (on line 62).

tobiasmo1 commented on pkg/models/commandvalue.go:
```diff
@@ -59,12 +61,20 @@ const (
+
+const (
+	// Policy limits should be located in go-mod-core-contract?
+	MaxBinaryBytes = 1048576
```
I added code comment and updated the constant to reflect **16MB** payload size limit.
